evol/problems/tsp.py

```diff
@@ -0,0 +1,74 @@
+import math
+from typing import List, Union
+
+from .problem import Problem
+
+
+def _rotating_window(arr):
+    for i, city in enumerate(arr):
+        yield arr[i-1], arr[i]
+
+
+class TSPProblem(Problem):
+    def __init__(self, distance_matrix):
+        self.distance_matrix = distance_matrix
+
+    @classmethod
+    def calc_distance_matrix(cls, coordinates: List[Union[tuple, list]]) -> List[List[Union[int, float]]]:
+        """
+        Creates a distance matrix from a list of city coordinates.
+        :param coordinates: An iterable that contains tuples or lists representing a x,y coordinate.
+        :return: A list of lists containing the distances between cities.
+        """
+        res = [[0 for i in coordinates] for j in coordinates]
+        for i, coord_i in enumerate(coordinates):
+            for j, coord_j in enumerate(coordinates):
+                dist = math.sqrt(sum([(z[0] - z[1])**2 for z in zip(coord_i, coord_j)]))
+                res[i][j] = dist
+                res[j][i] = dist
+        return res
+
+    def check_solution(self, solution: List[int]):
+        """
+        Check if the solution for the TSP problem is valid.
+        :param solution: List of integers which refer to cities.
+        :return: None, unless errors are raised.
+        """
+        set_solution = set(solution)
+        set_problem = set(range(len(self.distance_matrix)))
+        if len(solution) > len(self.distance_matrix):
+            raise RuntimeError("Solution is longer than number of towns!")
+        if set_solution != set_problem:
+            raise RuntimeError(f"Not all towns are visited! Am missing {set_problem.difference(set_solution)}")
+
+    def eval_function(self, solution: List[int]) -> Union[float, int]:
+        """
+        Calculates the cost of the current solution for the TSP problem.
+        :param solution: List of integers which refer to cities.
+        :return:
+        """
+        self.check_solution(solution=solution)
+        cost = 0
+        for t1, t2 in _rotating_window(solution):
+            cost += self.distance_matrix[t1][t2]
+        return cost
+
+
+_usa_capital_coords = [(-86.279118, 32.361538), (-134.419740, 58.301935), (-112.073844, 33.448457),
```

Comment:
Do we want to have csv files to be part of the library? It's not a biggy, but I figured having this in one file instead of two files might be clearer.

Reply:
After some thinking. It might make sense to have it into a separate file. Given a file in a format you could add a nice classmethod that allows the user to load in a larger file themselves as well. 

This direction cool?

Reply:
Exactly. We do need to make sure we package the data with evol, though. This would allow anyone to use the same class method with their data, which is nice.

Reply:
I am also wondering what the benefit is to adding `.csv` files containing the coordinates when we could also keep variables with coordinates.

Reply:
This is my current proposal. 

![image](https://user-images.githubusercontent.com/1019791/46903390-63f21c00-ced4-11e8-91df-2f140cd2fa37.png)

Reply:
In this case I guess it does not matter much, but when the `coordinates` module is loaded all data will be loaded into memory. If we have multiple of these datasets then I'd say that is a waste of memory.

Reply:
i imagine we keep the datasets limited but having a few is a good starting point. USA cities seems fine, not too big.

my main concern is the structure though.

Reply:
![image](https://user-images.githubusercontent.com/1019791/48303463-bee65580-e50a-11e8-9496-5df6fff5ad3d.png)

this is the current setup for all problems. i prefer to have the problems be accesible from the `__init__.py` part of the submodule. i've also added `functions` to demonstrate how other types of problems could be added.